frontend/catalyst/from_plxpr/decompose.py

```diff
@@ -0,0 +1,325 @@
+        alt_decomps (dict or None): A dictionary of alternative decomposition rules
+            to use in the decomposition graph.
+
+    Raises:
+        TypeError: if graph-based decomposition is not enabled.
+    """
+
+    # A mapping from operation names to the number of wires they act on.
+    # This is used when the operation is not in the captured operations
+    # but we still need to create a decomposition rule for it.
+    compiler_ops_num_wires: dict[str, int] = {
+        "CNOT": 2,
+        "ControlledPhaseShift": 2,
+        "CRot": 2,
+        "CRX": 2,
+        "CRY": 2,
+        "CRZ": 2,
+        "CSWAP": 3,
+        "CY": 2,
+        "CZ": 2,
+        "Hadamard": 1,
+        "Identity": 1,
+        "IsingXX": 2,
+        "IsingXY": 2,
+        "IsingYY": 2,
+        "IsingZZ": 2,
+        "SingleExcitation": 2,
+        "SingleExcitationPlus": 2,
+        "SingleExcitationMinus": 2,
+        "DoubleExcitation": 4,
+        "DoubleExcitationPlus": 4,
+        "DoubleExcitationMinus": 4,
+        "ISWAP": 2,
+        "PauliX": 1,
+        "PauliY": 1,
+        "PauliZ": 1,
+        "PhaseShift": 1,
+        "PSWAP": 2,
+        "Rot": 1,
+        "RX": 1,
+        "RY": 1,
+        "RZ": 1,
+        "S": 1,
+        "SWAP": 2,
+        "T": 1,
+        "Toffoli": 3,
+        "U1": 1,
+        "U2": 1,
+        "U3": 1,
+        "MultiRZ": -1,  # variable number of wires
+        "GlobalPhase": -1,  # variable number of wires
+    }
+
+    def __init__(
+        self,
+        *,
+        gate_set=None,
+        fixed_decomps=None,
+        alt_decomps=None,
+    ):  # pylint: disable=too-many-arguments
+
+        if not qml.decomposition.enabled_graph():  # pragma: no cover
+            raise TypeError(
+                "The GraphSolutionInterpreter can only be used when"
+                "graph-based decomposition is enabled."
+            )
+
+        self._gate_set = gate_set
+        self._fixed_decomps = fixed_decomps
+        self._alt_decomps = alt_decomps
+
+        self._captured = False
+        self._operations = set()
+        self._decomp_graph_solution = {}
+
+    def update_operations(self, operations):
+        """Update the set of captured operations.
+
+        Args:
+            operations (set): a set of pennylane operator instances
+        """
+        for op in operations:
+            # TODO: Although we deal with those ops not in compiler_ops_num_wires in the
+            # compiler-specific decomposition step, we should ideally have a way to specify
+            # the list of ops in the structured rule and their corresponding number of wires
+            # to solve the graph for them.
+            if op.name in self.compiler_ops_num_wires.keys():
+                self._operations.add(op)
+            else:
+                try:
+                    with qml.capture.pause():
+                        ops = op.decomposition()
+                        self.update_operations(ops)
```

Comment:
Let me know if I understand correctly: this `GraphSolutionInterpreter` is applied on the original `jaxpr` before the first stage decomposition, but the purpose of this interpreter is to build and solve the graph as a preparation for the second stage. This is why we only collect operations that are already in the compiler specific gate set, and if we encounter an operator that is not part of the compiler-specific gate set, we decompose it (with the old system?), and then update our record with its decomposition? Why not use this `GraphSolutionInterpreter` after you've already completed the first stage?